components/twitch/sources/common.js

```diff
@@ -0,0 +1,43 @@
+const twitch = require("../twitch.app.js");
+
+module.exports = {
+  dedupe: "unique",
+  props: {
+    twitch,
+    db: "$.service.db",
+  },
+  methods: {
+    async paginate(resourceFn, params, type, max = null) {
+      const items = [];
+      let done = false;
+      let count = 0;
+      while (!done) {
+        const { data, pagination } = await resourceFn(params);
+        for (const item of data) {
+          // if called from webhook component, get topics to return
+          if (type == "webhook") {
+            items.push(this.getTopicString(item));
+          }
+          // if called from polling component, emit data
+          else if (type == "polling") {
+            this.$emit(item, this.getMeta(item));
+          }
+          // if type is list, return list of items
+          else if (type == "list") {
+            items.push(item);
+          }
```

Comment:
If we transfor `paginate` into a generator, we can move the responsibility of figuring out what to do with the response of the API requests to the consumers of this function. Otherwise this function has to know the details of each consumer so that it can handle the API responses in the particular manner that the consumers wish.

Instead I propose the following:
```js
async function* paginate(resourceFn, params, max = null) {
  let done = false;
  let count = 0;
  do {
    const {
      data,
      pagination,
    } = await resourceFn(params);

    for (const item of data) {
      count++;
      if (max && count > max) {
        return;
      }

      yield item;
    }

    const { cursor } = pagination;
    params.after = cursor;
    done = !cursor;
  } while (!done);
}
```

Then each consumer can do this:
```js
for (const followed of followedUsers) {
  const params = {
    user_id: followed.to_id,
    // Period during which the video was created. Valid values: "all",
    // "day", "week", "month".
    period: "day",
  };
  const videosScan = await this.paginate(
    this.twitch.getVideos.bind(this),
    params,
    this.max,
  );
  for await (const video of videosScan) {
    const meta = this.getMeta(video);
    this.$emit(video, meta);
  }
}
```

Reply:
Updated the paginate method to be a generator function.